File: LeetCode/LeetCode/leetcode010.cs
using System;
using System.Collections;

// https://leetcode.com/problems/regular-expression-matching/

// parse
// http://blog.csdn.net/fzzying3/article/details/42057935
// http://meetqun.com/thread-2688-1-1.html

namespace leetcode010 {
	public class Solution {
		public bool IsMatch(string s, string p) {
			int sLen = s.Length;
			int pLen = p.Length;

			if (pLen == 0) {
				if (sLen == 0) {
					return true;
				}
				return false;
			}

			char[] pArr = p.ToCharArray ();
			char[] sArr = s.ToCharArray ();

			int sIdx = 0;
			if (pLen > 1 && pArr [1] == '*') {
				while (sIdx < sLen && (sArr [sIdx] == pArr [0] || pArr [0] == '.')) {
					if (IsMatch(s.Substring(sIdx), p.Substring(2))) {
						return true;
					}
					++sIdx;
				}
				return IsMatch (s.Substring(sIdx), p.Substring (2));
			}

			if (sIdx >= sLen) {
				return false;
			}

			return (sArr [sIdx] == pArr [0] || pArr [0] == '.') && IsMatch (s.Substring (1), p.Substring (1));
		}
	}

//	class MainClass
//	{
//		public static void Main (string[] args)
//		{
//			Solution s = new Solution ();
//			Console.WriteLine (s.IsMatch("aa", "a"));
//		}
//	}

}

File: LeetCode/LeetCode_CPP/leetcode010.cpp
// https://leetcode.com/problems/regular-expression-matching/

// http://www.cnblogs.com/grandyang/p/4461713.html

#include <iostream>
#include <string>
#include <math.h>

using namespace std;

class Solution {
public:
    /*
        (1)p[j+1]不是'*'。情况比较简单，只要判断当前s的i和p的j上的字符是否一样（如果有p在j上的字符是'.',也是相同），如果不同，返回false，否则，递归下一层i+1，j+1; 
        (2)p[j+1]是'*'。那么此时看从s[i]开始的子串，假设s[i],s[i+1],...s[i+k]都等于p[j]那么意味着这些都有可能是合适的匹配，那么递归对于剩下的(i,j+2),(i+1,j+2),...,(i+k,j+2)都要尝试（j+2是因为跳过当前和下一个'*'字符）。 
    */
    bool isMatch(string s, string p) {
        int slen = s.length(), plen = p.length();
        if (plen == 0 && slen == 0)
        {
            return true;
        }
        
        if (plen == 1)
        {
            if (slen == 1 && (s[0] == p[0] || p[0] == '.'))
            {
                return true;
            }
            return false;
        }

        if (p[1] != '*')
        {
            if (slen == 0)
            {
                return false;
            }
            return (s[0] == p[0] || p[0] == '.') && isMatch(s.substr(1), p.substr(1));
        }

        while(slen > 0 && (s[0] == p[0] || p[0] == '.'))
        {
            if (isMatch(s, p.substr(2)))
            {
                return true;
            }
            s = s.substr(1);
        }

        return isMatch(s, p.substr(2));
    }
};

int main()
{
    Solution s;

    cout << s.isMatch("ab", ".*c") << endl;

    return 0;
}
